### DISCUSSION TITLE
Can we get type safety for onEnter/onLeave

### DISCUSSION THREAD

**Discussion by nklhtv** [TanStack/router#1219] [Q&A] [answered]
The argument of [onEnter/onLeave](https://github.com/TanStack/router/blob/main/packages/react-router/src/route.ts#L218) is of type AnyRouteMatch which forces me to use type casting like:
```javascript
onEnter: (match) => {
    const loaderData = (match as RouteMatch<RegisteredRouter["routeTree"], typeof myRoute.types.id>).loaderData;
}
```
Is it safe to do type casting here?
Is the 'AnyRouteMatch' type used for reason here?
Can we use more explicit type here instead?

**Comment by ccapndave** [comment]
I'd also like to know about this.

**Comment by schiller-manuel** [comment, marked as answer]
I implemented this in https://github.com/TanStack/router/releases/tag/v1.29.0

**Comment by ccapndave** [comment]
Amazing!
